pkg/discovery/discovery.go

```diff
@@ -15,103 +15,334 @@ package discovery
+			initialClusterArgs[i] = fmt.Sprintf("%s=%s", address.Name, address.Address)
+		}
+
+		return "--initial-cluster=" + strings.Join(initialClusterArgs, ","), nil
 	}
-	tc, err := td.tcGetFn(ns, tcName)
-	if err != nil {
+
+}
+
+func validateEmpty(pdName PDName, clusterID ClusterID, pdURL url.URL) error {
+	if pdURL.String() == "" {
+		return fmt.Errorf("url is empty")
+	}
+	if pdName == "" {
+		return fmt.Errorf("pd name is empty")
+	}
+	if clusterID == "" {
+		return fmt.Errorf("cluster id is empty")
+	}
+	return nil
+}
+
+// Discover waits for all PD to join before starting the cluster
+// this approach was probably more useful before the PD isinitialized status API was available
+func (td *tidbDiscoveryMembers) Discover(pdName PDName, clusterID ClusterID, pdURL url.URL) (string, error) {
+	if err := validateEmpty(pdName, clusterID, pdURL); err != nil {
 		return "", err
 	}
-	keyName := fmt.Sprintf("%s/%s", ns, tcName)
-	// TODO: the replicas should be the total replicas of pd sets.
-	replicas := tc.Spec.PD.Replicas
+	glog.Infof("url is: %s", pdURL.String())
 
-	currentCluster := td.clusters[keyName]
-	if currentCluster == nil || currentCluster.resourceVersion != tc.ResourceVersion {
-		td.clusters[keyName] = &clusterInfo{
-			resourceVersion: tc.ResourceVersion,
-			peers:           map[string]struct{}{},
+	cluster, gerr := td.refresh.GetCluster(string(clusterID))
+	if gerr != nil {
+		return "", gerr
+	}
+
+	td.lock.Lock()
+	defer td.lock.Unlock()
+
+	currentCluster := td.clusters[string(clusterID)]
+	if currentCluster == nil || currentCluster.resourceVersion != cluster.ResourceVersion {
+		currentCluster = &clusterInfo{
+			resourceVersion: cluster.ResourceVersion,
+			peers:           make(map[PDName]url.URL),
 		}
+		td.clusters[string(clusterID)] = currentCluster
 	}
-	currentCluster = td.clusters[keyName]
-	currentCluster.peers[podName] = struct{}{}
+	currentCluster.peers[pdName] = pdURL
+
+	if len(currentCluster.peers) == int(cluster.Replicas) {
+		delete(currentCluster.peers, pdName)
 
-	if len(currentCluster.peers) == int(replicas) {
-		delete(currentCluster.peers, podName)
-		return fmt.Sprintf("--initial-cluster=%s=%s://%s", podName, tc.Scheme(), advertisePeerUrl), nil
+		if pdURL.Scheme == "" && cluster.Scheme != "" {
+			pdURL.Scheme = cluster.Scheme
+		}
+
+		return fmt.Sprintf("--initial-cluster=%s=%s", pdName, pdURL.String()), nil
 	}
 
-	pdClient := td.pdControl.GetPDClient(pdapi.Namespace(tc.GetNamespace()), tc.GetName(), tc.Spec.EnableTLSCluster)
-	membersInfo, err := pdClient.GetMembers()
+	addresses, err := td.GetClientAddresses(clusterID)
 	if err != nil {
 		return "", err
 	}
+	delete(currentCluster.peers, pdName)
+	return fmt.Sprintf("--join=%s", strings.Join(addresses, ",")), nil
+}
+
+// ParseURL calls url.Parse but parses according to user expecdtation when there is no protocol or leading "//" or host brackets "[]"
+func ParseURL(inputURL string) (url.URL, error) {
+	parsedURL, err := url.Parse(inputURL)
```

Comment:
I agree that normalization would be easier. I should document that this function preserves the input rather than normalizing it so that when one does `url.String()` they get back the original url. Currently this behavior is desirable since TiKV doesn't want a protocol prefix. However, I should probably fix this issue elsewhere.

Reply:
The current behavior is much better though if one doesn't want to assume http. I am changing the name to `ParseAddress`.